compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs

```diff
@@ -56,15 +65,26 @@ decodeString :: TL.Text -> T.Text
 decodeString = TL.toStrict
 
 -- | Decode a string that will be interned in DAML-LF 1.7 and onwards.
-decodeInternableString :: TL.Text -> Decode T.Text
-decodeInternableString = pure . decodeString
+decodeInternableStrings :: V.Vector TL.Text -> Int32 -> Decode [T.Text]
+decodeInternableStrings strs id
+    | V.null strs = lookupStringList id
+    | id == 0 = pure $ map decodeString (V.toList strs)
+    | otherwise = throwError $ ParseError "items and interned id both set for string list"
 
 -- | Decode the name of a syntactic object, e.g., a variable or a data
 -- constructor. These strings are mangled to escape special characters. All
 -- names will be interned in DAML-LF 1.7 and onwards.
-decodeName :: (T.Text -> a) -> TL.Text -> Decode a
-decodeName wrapName mangled = do
-    mangled <- decodeInternableString mangled
+decodeName
+    :: Util.EitherLike m1 m2 m3 m4 m5 TL.Text Int32 e
```

Comment:
`EitherLike` scares me, can we not just pass in a function `e -> Either TL.Text Int32`?

Reply:
or even better, just fmap at the usesite. and pass in `Maybe (Either TL.Text Int32)` here.

Reply:
Or make a custom typeclass with a method `e -> Either TL.Text Int32` and convert to that if you really don’t want to pass this explicitly. You can still derive the implementation using generics but at least you don’t leak 5 unused type variables to every use site.

Reply:
I went with the last suggestion. Thanks for pushing me to improve this.